src/ast/codegen_helper.h

```diff
@@ -10,22 +10,27 @@ inline bool needMemcpy(const SizedType &stype)
   return stype.IsAggregate() || stype.IsTimestampTy() || stype.IsCgroupPathTy();
 }
 
-inline bool shouldBeOnStackAlready(const SizedType &type)
+// BPF memory is memory that the program can access with a regular
+// dereference. This could mean the value is on the stack, a map, or
+// maybe something else (like BPF arenas) in the future.
+//
+// The necessarily means that a bpf_probe_read_*() is _NOT_ required.
+inline bool shouldBeInBpfMemoryAlready(const SizedType &type)
 {
   return type.IsStringTy() || type.IsBufferTy() || type.IsInetTy() ||
          type.IsUsymTy() || type.IsKstackTy() || type.IsUstackTy() ||
          type.IsTupleTy() || type.IsTimestampTy() || type.IsMacAddressTy() ||
          type.IsCgroupPathTy();
 }
 
-inline bool onStack(const SizedType &type)
+inline bool inBpfMemory(const SizedType &type)
 {
-  return type.is_internal || shouldBeOnStackAlready(type);
+  return type.is_internal || shouldBeInBpfMemoryAlready(type);
```

Comment:
The next thing to clear up is this 'is_internal' flag. I don't understand why it only applies to `Type::inet` and `Type::mac_address`. But not for this PR.

Reply:
Yeah, there's probably some stuff we should clean up around this. It was probably written around the time when we had a pretty sharp transition in people working on this project, and the new guys (me and Viktor) were just figuring stuff out as we went

Reply:
Definitely some messiness with this.

`is_internal` is only hardcoded to true for `inet` and `mac_address` (I don't know why they don't just use `shouldBeInBpfMemoryAlready()`), but it's also set dynamically for other types, e.g. structs when assigned to a map: https://github.com/bpftrace/bpftrace/blob/898ae5e10a3e5f5c14290623b21507da776348cc/src/ast/passes/semantic_analyser.cpp#L2532

Reply:
I'll try to make some time to clean it up.

Reply:
Maybe we could move across to just using the address space markers? `AddrSpace::bpf` means `is_internal` / `inBpfMemory()` and `AddrSpace::kernel` or `AddrSpace::user` otherwise

Reply:
That makes sense to me. And it probably opens up some other interesting optimizations in the future. I bet we're emitting bpf_probe_read_kernel()s where they aren't strictly necessary at the moment.